### DISCUSSION TITLE
How should RouterMiddleware be used?

### DISCUSSION THREAD

**Discussion by finlaydotb** [oakserver/oak#497] [Q&A] [answered]
In an application using an older version of oak, the middleware get's plugged into the handler as follows:

```
router.get("/path", isAdmin, getHandler)
```

Where `isAdmin` is something like:

```
export const isAdmin = async (ctx: RouterContext<"/path">, next: Function) => {
    await next();
}
```

This no longer compiles. On further investigation I discovered it seems the middleware should now be typed as `RouterMiddleware`

But when I updated `isAdmin` to use `RouterMiddleware` like this:

```
export const isAdmin = async (ctx: RouterMiddleware<"/path">, next: Function) => {
    await next();
}
```

It now fails to compile with the following error:

```
Argument of type '(ctx: RouterMiddleware</path>, next: Function) => Promise<void>' is not assignable to parameter of type 'RouterMiddleware<"/path", Record<string | number, string | undefined>, Record<string, any>>'.
```

Any ideas on how to fix this?

**Comment by kitsonk** [comment, marked as answer]
> But when I updated `isAdmin` to use `RouterMiddleware` like this:
> 
> ```
> export const isAdmin = async (ctx: RouterMiddleware<"/path">, next: Function) => {
>     await next();
> }
> ```

Try:

```ts
export const isAdmin: RouterMiddleware<"/path"> = async (ctx, next) => {
    await next();
}
```
